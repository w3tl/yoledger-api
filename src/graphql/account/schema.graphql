extend type Query {
  accounts(type: AccountType!): [Account]
  account(id: ID!): Account
}

extend type Mutation {
  addAccount(input: AddAccountInput!): AddAccountPayload
}

type Account {
  id: ID!
  name: String!
  balance: Float
  type: AccountType!
}

enum AccountType {
  ASSET
  EXPENSE
  INCOME
  VENDOR
}

input AddAccountInput {
  name: String!
  balance: Float
  type: AccountType!
}

type AddAccountPayload {
  account: Account
}
