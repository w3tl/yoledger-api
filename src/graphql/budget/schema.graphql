extend type Query {
  budgets(dateStart: Date!, dateEnd: Date!): Budgets
  budget(account: String!, dateStart: Date!, dateEnd: Date!): [Budget]
}
extend type Mutation {
  upsertBudget(input: UpsertBudgetInput!): UpsertBudgetPayload
}

type Budgets {
  accounts: [Account]
}

type Budget {
  id: ID!
  date: Date!
  account: Account!
  amount: Float
}

input UpsertBudgetInput {
  account: String!
  date: Date!
  amount: Float!
}

type UpsertBudgetPayload {
  success: Boolean
  budget: Budget
}
