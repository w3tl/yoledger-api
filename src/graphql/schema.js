import { Schema as userTypeDefs } from './user';
import { Schema as transactionTypeDefs } from './transaction';
import { Schema as accountTypeDefs } from './account';
import { Schema as budgetTypeDefs } from './budget';

export default `
type Query {
  dummy: Boolean
}
type Mutation {
  dummy: Boolean
}
scalar Date
${userTypeDefs}
${accountTypeDefs}
${transactionTypeDefs}
${budgetTypeDefs}
`;
