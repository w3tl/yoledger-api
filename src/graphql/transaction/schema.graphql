extend type Query {
  transactions(dateStart: Date!, dateEnd: Date, page: Int = 0, itemsPerPage: Int = 20): [Transaction]
  transaction(id: ID!): Transaction
}
extend type Mutation {
  addTransaction(input: AddTransactionInput!): AddTransactionPayload
  deleteTransaction(id: ID!): DeleteTransactionPayload
  updateTransaction(id: ID!, input: UpdateTransactionInput!): UpdateTransactionPayload
}

type Transaction {
  id: ID!
  amount: Float!
  source: Account!
  destination: Account!
  date: Date!
}
input AddTransactionInput {
  source: String!
  destination: String!
  amount: Float!
  date: Date
}
type AddTransactionPayload {
  transaction: Transaction
}
type DeleteTransactionPayload {
  success: Boolean!
  id: ID
}
input UpdateTransactionInput {
  amount: Float
  source: String
  destination: String
  date: Date
}
type UpdateTransactionPayload {
  transaction: Transaction
}
