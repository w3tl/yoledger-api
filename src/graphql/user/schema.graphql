extend type Query {
  user: User
}

extend type Mutation {
  signin(login: String!, password: String!): SignupPayload!
  signout(token: String): SignoutPayload!
  changePassword(oldPassword: String!, newPassword: String!): ChangePasswordPayload!
}

type User {
  username: String
}

type SignupPayload {
  token: String!
}

type SignoutPayload {
  success: Boolean
}

type ChangePasswordPayload {
  token: String
}
