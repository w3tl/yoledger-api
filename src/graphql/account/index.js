import Schema from './schema.graphql';
import { Query, Mutation, Account } from './resolver';

export {
  Schema, Query, Mutation, Account,
};
