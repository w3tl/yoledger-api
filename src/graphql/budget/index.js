import Schema from './schema.graphql';
import {
  Query, Mutation, Budget,
} from './resolver';

export {
  Schema, Query, Mutation, Budget,
};
