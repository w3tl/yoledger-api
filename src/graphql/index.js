import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './schema';
import resolvers from './resolvers';
import dataloaders from './dataloaders';
import {
  User,
  Account,
  Transaction,
  Budget,
} from '../model';

export default makeExecutableSchema({
  typeDefs,
  resolvers,
});

export const models = (connection, userId) => ({
  userModel: new User(connection.db(), userId),
  accountModel: new Account(connection.db(), userId),
  transactionModel: new Transaction(connection.db(), userId),
  budgetModel: new Budget(connection.db(), userId),
});

export { dataloaders };
