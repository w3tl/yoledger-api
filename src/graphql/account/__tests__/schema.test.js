jest.mock('mongodb');

import { graphql } from 'graphql';
import schema from '../../index';
import { Account } from '../../../model';

describe('account schema', () => {
  const { connection } = require('mongodb');
  const accountModel = new Account(connection.db(), 'admin');

  it('Should be return accounts', async () => {
    const query = `
    query accounts($type: AccountType!) {
      accounts(type: $type) {
         name
         balance
      }
    }
    `;

    const rootValue = {};
    const context = { models: { accountModel } };
    const variables = {
      type: 'ASSET',
    };

    const result = await graphql(schema, query, rootValue, context, variables);
    const { data } = result;

    expect(data.accounts).toMatchSnapshot('query ASSET accounts');
  });
});
