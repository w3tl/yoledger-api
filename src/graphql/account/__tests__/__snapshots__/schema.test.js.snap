// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`account schema Should be return accounts: query ASSET accounts 1`] = `
Array [
  Object {
    "balance": 1500.5,
    "name": "Bank card",
  },
  Object {
    "balance": 143,
    "name": "Cash",
  },
]
`;
